test/functional/feature_reindex.py

```diff
@@ -20,18 +25,53 @@ def set_test_params(self):
 
     def reindex(self, justchainstate=False):
         self.nodes[0].generatetoaddress(3, self.nodes[0].get_deterministic_priv_key().address)
+        self.sync_blocks()
         blockcount = self.nodes[0].getblockcount()
         self.stop_nodes()
         extra_args = [["-reindex-chainstate" if justchainstate else "-reindex"]]
         self.start_nodes(extra_args)
         wait_until(lambda: self.nodes[0].getblockcount() == blockcount)
-        self.log.info("Success")
+
+    # Check that blocks can be processed out of order
+    def out_of_order(self):
+        # The previous test created 12 blocks
+        assert_equal(self.nodes[0].getblockcount(), 12)
+        self.stop_nodes()
+
+        # In this test environment, blocks will always be in order (since
+        # we're generating them rather than getting them from peers, so to
+        # test out-of-order handling, swap blocks 1 and 2 on disk.
+        blk0 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00000.dat')
+        with open(blk0, 'r+b') as bf:
+            b = bf.read(814)
+
+            # This is really just checking the test; these are is the first
+            # "marker" bytes of the four blocks, genesis and 3 more
+            # (these offsets are extremely unlikely ever to change)
+            assert_equal(b[0], 0xfa)
+            assert_equal(b[293], 0xfa)
+            assert_equal(b[553], 0xfa)
+            assert_equal(b[813], 0xfa)
+
+            # Swap the same-size second and third blocks (don't change genesis).
+            bf.seek(293)
+            bf.write(b[553:813])
+            bf.write(b[293:553])
+
+        extra_args = [["-reindex", "-debug=reindex"]]
+        self.start_nodes(extra_args)
+        wait_until(lambda: self.nodes[0].getblockcount() > 0)
+
+        # All blocks should be accepted and processed.
+        assert_equal(self.nodes[0].getblockcount(), 12)
```

Comment:
I did verify that if I break this PR's code, this test doesn't pass. I know hardcoded constants are generally bad, but I think they're okay here, because the size of an empty `regtest` block is very unlikely to change, and if it does, this test will fail, and it should be pretty easy for someone to figure out why and fix the test. (That is, the test will not falsely succeed.)